python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+            self.size_frame_meta,
+            self.size_frame_data,
+        ) = self._read_structural_data()
+
+        if self.num_writes == self.invalid_value:
+            raise SharedMemoryError(
+                "The shared memory referenced by '{}' was marked as invalid by the writer.".format(
+                    self.shared_mem_name
+                )
+            )
+
+        self.len_data = self.num_freq * self.num_time
+
+        self.size_access_record = self.len_data * self.size_access_record_entry
+        self.pos_access_record = self.size_structural_data
+
+        self.size_data = self.len_data * self.size_frame
+        self.pos_data = self.pos_access_record + self.size_access_record
+
+        self._initial_validation()
+
+        os.close(shared_mem.fd)
+
+        self._data = np.ndarray(
+            (buffer_size, self.num_freq), dtype=np.dtype((np.void, self.size_frame))
+        )
+        logger.debug(
+            "Created buffer for copy of data (size: {})".format(self._data.shape)
+        )
+        self._last_access_record = None
+
+    def _initial_validation(self):
+        shared_mem_size = (
+            self.size_structural_data + self.size_access_record + self.size_data
+        )
+        if shared_mem_size != self.shared_mem.size():
+            raise SharedMemoryError(
+                "Expected shared memory to have size {} (but has {}).".format(
+                    shared_mem_size, self.shared_mem.size()
+                )
+            )
+
+    def _read_structural_data(self):
+        num_writes = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        num_time = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        num_freq = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_frame = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_meta = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        size_data = struct.unpack_from(
+            FMT_UINT64_T, self.shared_mem.read(SIZE_UINT64_T)
+        )[0]
+        return (
+            num_writes,
+            num_time,
+            num_freq,
+            size_frame,
+            size_meta,
+            size_data,
+        )
+
+    def __del__(self):
+        self.semaphore.release()
+        try:
+            self.semaphore.unlink()
+        except posix_ipc.ExistentialError:
+            logger.debug("Semaphore file did not exist when trying to unlink from it.")
+
+        self.shared_mem.close()
+        try:
+            posix_ipc.unlink_shared_memory(self.shared_mem_name)
+        except posix_ipc.ExistentialError:
+            logger.debug(
+                "Shared memory file did not exist when trying to unlink from it."
+            )
+
+    def read_last(self, n: int):
```

Comment:
This code needs to run in the cal broker which is Python 2. So no type annotations I'm afraid.